elastic/logstash
# Pipeline to pipeline communication acked queue improvements.

## Release notes

Resolves upstream pipeline termination when using pipeline to pipeline communication if writing to a closed queue.

## What does this PR do?
Adresses number of Pipeline to Pipeline upstream issues including abnormal termination issues:
  - `"Tried to write to a closed queue." IllegalStateException` exception, terminates upstream pipeline
    Resolution: ideally locking `input` in the pipeline bus while sending the inflight event would be a synchronized task to prevent such cases and we have [this option](https://gist.github.com/yaauie/5bf4db9f7f7bd17d6b0b84741940a9a1) (big thanks @yaauie) to consider. However, blocking mechanism would be an expensive and expected behavior is to catch exception (failed inflight send position) and let process retry as we have a logic in pipeline bus.

  - Inflight events duplicate while sending
    Issue: when sending inflight events to terminated downstream pipeline, we retry from the failed position we lastly sent. However, we have a bug that we copy the events to send from the point where cloned event's failed position.
    Resolution: the failed position should be a cumulative number, should correspond origin event's position as `lastFailedPosition += lastResponse.getSequencePosition();`

  - `"Cannot invoke "org.logstash.ackedqueue.Page.hasCapacity(int)" because "this.headPage" is null"`
      `NullPointerException` exception. It happens when `ensure_delivery => true`, the situation where queue is closed but input is trying to write into it.
    (reproduce with 3 upstream pipeline workers, single downstream pipeline with 1000 batch)
    Resolution: use atomic `Queue->isClosed()` and check for null to prevent headPage NPE
    Note: when `ensure_delivery => false`, pipeline ignores failed inflight events, means it doesn't write to queue.

  - Queue closed check improvements
    `JRubyWrappedAckedQueueExt` implemented its own `isClosed` atomic variable to identify if queue is closed. However, since `input` is a volatile in pipeline bus (even if not), aligned on actual `closed` variable of original `Queue` is an accurate.


## Why is it important/What is the impact to the user?



## Checklist




## Author's Checklist


## How to test this PR locally
- See https://github.com/elastic/logstash/issues/15159 how to test `Tried to write to a closed queue.` the issue.
- Inflight events lost issue is hard to test.
- `Cannot invoke "org.logstash.ackedqueue.Page.hasCapacity(int)" because "this.headPage" is null` NPE test also hard. There is 1% percent of possibility to make it happen when using multiple upstream workers and single downstream worker with small (ex: up to 100) batch size.

## Related issues

- Closes #15159 

## Use cases

Please see the description for the use-case and details of resolution

## Screenshots

## Logs after change, see issue description for before change logs.
```
[2023-08-02T11:51:29,367][INFO ][logstash.pipelineaction.reload] Reloading pipeline {"pipeline.id"=>:downstream}
[2023-08-02T11:51:33,814][INFO ][logstash.javapipeline    ][downstream] Pipeline terminated {"pipeline.id"=>"downstream"}
[2023-08-02T11:51:33,815][WARN ][org.logstash.plugins.pipeline.PipelineBus][upstream][c28fc23a0d8a4f050c8fd52237d2d161521029e2ebd918ed033ba8725cb5886a] Attempted to send events to 'myDownstreamPipeline' but that address reached error condition with 35 events remaining. Will Retry. Root cause Tried to write to a closed queue.
[2023-08-02T11:51:34,448][INFO ][logstash.javapipeline    ] Pipeline `downstream` is configured with `pipeline.ecs_compatibility: v8` setting. All plugins in this pipeline will default to `ecs_compatibility => v8` unless explicitly configured otherwise.
[2023-08-02T11:51:34,450][WARN ][logstash.javapipeline    ][downstream] 'pipeline.ordered' is enabled and is likely less efficient, consider disabling if preserving event order is not necessary
[2023-08-02T11:51:34,451][INFO ][logstash.javapipeline    ][downstream] Starting pipeline {:pipeline_id=>"downstream", "pipeline.workers"=>1, "pipeline.batch.size"=>100, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>100, "pipeline.sources"=>["central pipeline management"], :thread=>"#<Thread:0x2e3fbf20 /Users/mashhur/Dev/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:134 run>"}
[2023-08-02T11:51:34,459][INFO ][logstash.javapipeline    ][downstream] Pipeline Java execution initialization time {"seconds"=>0.01}
[2023-08-02T11:51:34,460][INFO ][logstash.javapipeline    ][downstream] Pipeline started {"pipeline.id"=>"downstream"}
[2023-08-02T11:51:34,466][INFO ][logstash.agent           ] Pipelines running {:count=>2, :running_pipelines=>[:downstream, :upstream], :non_running_pipelines=>[]}
```
